// @openrouter/sdk/esm/models/chatgenerationparams.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../lib/primitives.js";
import * as openEnums from "../types/enums.js";
import { ChatStreamOptions$outboundSchema, } from "./chatstreamoptions.js";
import { Message$outboundSchema, } from "./message.js";
import { ReasoningSummaryVerbosity$outboundSchema, } from "./reasoningsummaryverbosity.js";
import { ResponseFormatJSONSchema$outboundSchema, } from "./responseformatjsonschema.js";
import { ResponseFormatTextGrammar$outboundSchema, } from "./responseformattextgrammar.js";
import { Schema0$outboundSchema, } from "./schema0.js";
import { ToolDefinitionJson$outboundSchema, } from "./tooldefinitionjson.js";
export const ChatGenerationParamsDataCollection = {
    Deny: "deny",
    Allow: "allow",
};
export const Quantizations = {
    Int4: "int4",
    Int8: "int8",
    Fp4: "fp4",
    Fp6: "fp6",
    Fp8: "fp8",
    Fp16: "fp16",
    Bf16: "bf16",
    Fp32: "fp32",
    Unknown: "unknown",
};
export const Sort = {
    Price: "price",
    Throughput: "throughput",
    Latency: "latency",
};
export const ChatGenerationParamsPdfEngine = {
    MistralOcr: "mistral-ocr",
    PdfText: "pdf-text",
    Native: "native",
};
export const ChatGenerationParamsEngine = {
    Native: "native",
    Exa: "exa",
};
export const ChatGenerationParamsRoute = {
    Fallback: "fallback",
    Sort: "sort",
};
export const Effort = {
    None: "none",
    Minimal: "minimal",
    Low: "low",
    Medium: "medium",
    High: "high",
    Xhigh: "xhigh",
};
/** @internal */
export const ChatGenerationParamsDataCollection$outboundSchema = openEnums.outboundSchema(ChatGenerationParamsDataCollection);
/** @internal */
export const Quantizations$outboundSchema = openEnums.outboundSchema(Quantizations);
/** @internal */
export const Sort$outboundSchema = openEnums
    .outboundSchema(Sort);
/** @internal */
export const ChatGenerationParamsMaxPrice$outboundSchema = z.object({
    prompt: z.any().optional(),
    completion: z.any().optional(),
    image: z.any().optional(),
    audio: z.any().optional(),
    request: z.any().optional(),
});
export function chatGenerationParamsMaxPriceToJSON(chatGenerationParamsMaxPrice) {
    return JSON.stringify(ChatGenerationParamsMaxPrice$outboundSchema.parse(chatGenerationParamsMaxPrice));
}
/** @internal */
export const ChatGenerationParamsProvider$outboundSchema = z.object({
    allowFallbacks: z.nullable(z.boolean()).optional(),
    requireParameters: z.nullable(z.boolean()).optional(),
    dataCollection: z.nullable(ChatGenerationParamsDataCollection$outboundSchema)
        .optional(),
    zdr: z.nullable(z.boolean()).optional(),
    enforceDistillableText: z.nullable(z.boolean()).optional(),
    order: z.nullable(z.array(Schema0$outboundSchema)).optional(),
    only: z.nullable(z.array(Schema0$outboundSchema)).optional(),
    ignore: z.nullable(z.array(Schema0$outboundSchema)).optional(),
    quantizations: z.nullable(z.array(Quantizations$outboundSchema)).optional(),
    sort: z.nullable(Sort$outboundSchema).optional(),
    maxPrice: z.lazy(() => ChatGenerationParamsMaxPrice$outboundSchema)
        .optional(),
    minThroughput: z.nullable(z.number()).optional(),
    maxLatency: z.nullable(z.number()).optional(),
}).transform((v) => {
    return remap$(v, {
        allowFallbacks: "allow_fallbacks",
        requireParameters: "require_parameters",
        dataCollection: "data_collection",
        enforceDistillableText: "enforce_distillable_text",
        maxPrice: "max_price",
        minThroughput: "min_throughput",
        maxLatency: "max_latency",
    });
});
export function chatGenerationParamsProviderToJSON(chatGenerationParamsProvider) {
    return JSON.stringify(ChatGenerationParamsProvider$outboundSchema.parse(chatGenerationParamsProvider));
}
/** @internal */
export const ChatGenerationParamsPluginResponseHealing$outboundSchema = z.object({
    id: z.literal("response-healing"),
    enabled: z.boolean().optional(),
});
export function chatGenerationParamsPluginResponseHealingToJSON(chatGenerationParamsPluginResponseHealing) {
    return JSON.stringify(ChatGenerationParamsPluginResponseHealing$outboundSchema.parse(chatGenerationParamsPluginResponseHealing));
}
/** @internal */
export const ChatGenerationParamsPdfEngine$outboundSchema = openEnums.outboundSchema(ChatGenerationParamsPdfEngine);
/** @internal */
export const ChatGenerationParamsPdf$outboundSchema = z.object({
    engine: ChatGenerationParamsPdfEngine$outboundSchema.optional(),
});
export function chatGenerationParamsPdfToJSON(chatGenerationParamsPdf) {
    return JSON.stringify(ChatGenerationParamsPdf$outboundSchema.parse(chatGenerationParamsPdf));
}
/** @internal */
export const ChatGenerationParamsPluginFileParser$outboundSchema = z.object({
    id: z.literal("file-parser"),
    enabled: z.boolean().optional(),
    maxFiles: z.number().optional(),
    pdf: z.lazy(() => ChatGenerationParamsPdf$outboundSchema).optional(),
}).transform((v) => {
    return remap$(v, {
        maxFiles: "max_files",
    });
});
export function chatGenerationParamsPluginFileParserToJSON(chatGenerationParamsPluginFileParser) {
    return JSON.stringify(ChatGenerationParamsPluginFileParser$outboundSchema.parse(chatGenerationParamsPluginFileParser));
}
/** @internal */
export const ChatGenerationParamsEngine$outboundSchema = openEnums.outboundSchema(ChatGenerationParamsEngine);
/** @internal */
export const ChatGenerationParamsPluginWeb$outboundSchema = z.object({
    id: z.literal("web"),
    enabled: z.boolean().optional(),
    maxResults: z.number().optional(),
    searchPrompt: z.string().optional(),
    engine: ChatGenerationParamsEngine$outboundSchema.optional(),
}).transform((v) => {
    return remap$(v, {
        maxResults: "max_results",
        searchPrompt: "search_prompt",
    });
});
export function chatGenerationParamsPluginWebToJSON(chatGenerationParamsPluginWeb) {
    return JSON.stringify(ChatGenerationParamsPluginWeb$outboundSchema.parse(chatGenerationParamsPluginWeb));
}
/** @internal */
export const ChatGenerationParamsPluginModeration$outboundSchema = z.object({
    id: z.literal("moderation"),
});
export function chatGenerationParamsPluginModerationToJSON(chatGenerationParamsPluginModeration) {
    return JSON.stringify(ChatGenerationParamsPluginModeration$outboundSchema.parse(chatGenerationParamsPluginModeration));
}
/** @internal */
export const ChatGenerationParamsPluginUnion$outboundSchema = z.union([
    z.lazy(() => ChatGenerationParamsPluginModeration$outboundSchema),
    z.lazy(() => ChatGenerationParamsPluginWeb$outboundSchema),
    z.lazy(() => ChatGenerationParamsPluginFileParser$outboundSchema),
    z.lazy(() => ChatGenerationParamsPluginResponseHealing$outboundSchema),
]);
export function chatGenerationParamsPluginUnionToJSON(chatGenerationParamsPluginUnion) {
    return JSON.stringify(ChatGenerationParamsPluginUnion$outboundSchema.parse(chatGenerationParamsPluginUnion));
}
/** @internal */
export const ChatGenerationParamsRoute$outboundSchema = openEnums.outboundSchema(ChatGenerationParamsRoute);
/** @internal */
export const Effort$outboundSchema = openEnums
    .outboundSchema(Effort);
/** @internal */
export const Reasoning$outboundSchema = z.object({
    effort: z.nullable(Effort$outboundSchema).optional(),
    summary: z.nullable(ReasoningSummaryVerbosity$outboundSchema).optional(),
});
export function reasoningToJSON(reasoning) {
    return JSON.stringify(Reasoning$outboundSchema.parse(reasoning));
}
/** @internal */
export const ChatGenerationParamsResponseFormatPython$outboundSchema = z.object({
    type: z.literal("python"),
});
export function chatGenerationParamsResponseFormatPythonToJSON(chatGenerationParamsResponseFormatPython) {
    return JSON.stringify(ChatGenerationParamsResponseFormatPython$outboundSchema.parse(chatGenerationParamsResponseFormatPython));
}
/** @internal */
export const ChatGenerationParamsResponseFormatJSONObject$outboundSchema = z.object({
    type: z.literal("json_object"),
});
export function chatGenerationParamsResponseFormatJSONObjectToJSON(chatGenerationParamsResponseFormatJSONObject) {
    return JSON.stringify(ChatGenerationParamsResponseFormatJSONObject$outboundSchema.parse(chatGenerationParamsResponseFormatJSONObject));
}
/** @internal */
export const ChatGenerationParamsResponseFormatText$outboundSchema = z.object({
    type: z.literal("text"),
});
export function chatGenerationParamsResponseFormatTextToJSON(chatGenerationParamsResponseFormatText) {
    return JSON.stringify(ChatGenerationParamsResponseFormatText$outboundSchema.parse(chatGenerationParamsResponseFormatText));
}
/** @internal */
export const ChatGenerationParamsResponseFormatUnion$outboundSchema = z.union([
    z.lazy(() => ChatGenerationParamsResponseFormatText$outboundSchema),
    z.lazy(() => ChatGenerationParamsResponseFormatJSONObject$outboundSchema),
    ResponseFormatJSONSchema$outboundSchema,
    ResponseFormatTextGrammar$outboundSchema,
    z.lazy(() => ChatGenerationParamsResponseFormatPython$outboundSchema),
]);
export function chatGenerationParamsResponseFormatUnionToJSON(chatGenerationParamsResponseFormatUnion) {
    return JSON.stringify(ChatGenerationParamsResponseFormatUnion$outboundSchema.parse(chatGenerationParamsResponseFormatUnion));
}
/** @internal */
export const ChatGenerationParamsStop$outboundSchema = z.union([z.string(), z.array(z.string())]);
export function chatGenerationParamsStopToJSON(chatGenerationParamsStop) {
    return JSON.stringify(ChatGenerationParamsStop$outboundSchema.parse(chatGenerationParamsStop));
}
/** @internal */
export const Debug$outboundSchema = z.object({
    echoUpstreamBody: z.boolean().optional(),
}).transform((v) => {
    return remap$(v, {
        echoUpstreamBody: "echo_upstream_body",
    });
});
export function debugToJSON(debug) {
    return JSON.stringify(Debug$outboundSchema.parse(debug));
}
/** @internal */
export const ChatGenerationParams$outboundSchema = z.object({
    provider: z.nullable(z.lazy(() => ChatGenerationParamsProvider$outboundSchema)).optional(),
    plugins: z.array(z.union([
        z.lazy(() => ChatGenerationParamsPluginModeration$outboundSchema),
        z.lazy(() => ChatGenerationParamsPluginWeb$outboundSchema),
        z.lazy(() => ChatGenerationParamsPluginFileParser$outboundSchema),
        z.lazy(() => ChatGenerationParamsPluginResponseHealing$outboundSchema),
    ])).optional(),
    route: z.nullable(ChatGenerationParamsRoute$outboundSchema).optional(),
    user: z.string().optional(),
    sessionId: z.string().optional(),
    messages: z.array(Message$outboundSchema),
    model: z.string().optional(),
    models: z.array(z.string()).optional(),
    frequencyPenalty: z.nullable(z.number()).optional(),
    logitBias: z.nullable(z.record(z.string(), z.number())).optional(),
    logprobs: z.nullable(z.boolean()).optional(),
    topLogprobs: z.nullable(z.number()).optional(),
    maxCompletionTokens: z.nullable(z.number()).optional(),
    maxTokens: z.nullable(z.number()).optional(),
    metadata: z.record(z.string(), z.string()).optional(),
    presencePenalty: z.nullable(z.number()).optional(),
    reasoning: z.lazy(() => Reasoning$outboundSchema).optional(),
    responseFormat: z.union([
        z.lazy(() => ChatGenerationParamsResponseFormatText$outboundSchema),
        z.lazy(() => ChatGenerationParamsResponseFormatJSONObject$outboundSchema),
        ResponseFormatJSONSchema$outboundSchema,
        ResponseFormatTextGrammar$outboundSchema,
        z.lazy(() => ChatGenerationParamsResponseFormatPython$outboundSchema),
    ]).optional(),
    seed: z.nullable(z.int()).optional(),
    stop: z.nullable(z.union([z.string(), z.array(z.string())])).optional(),
    stream: z.boolean().default(false),
    streamOptions: z.nullable(ChatStreamOptions$outboundSchema).optional(),
    temperature: z.nullable(z.number()).optional(),
    toolChoice: z.any().optional(),
    tools: z.array(ToolDefinitionJson$outboundSchema).optional(),
    topP: z.nullable(z.number()).optional(),
    debug: z.lazy(() => Debug$outboundSchema).optional(),
}).transform((v) => {
    return remap$(v, {
        sessionId: "session_id",
        frequencyPenalty: "frequency_penalty",
        logitBias: "logit_bias",
        topLogprobs: "top_logprobs",
        maxCompletionTokens: "max_completion_tokens",
        maxTokens: "max_tokens",
        presencePenalty: "presence_penalty",
        responseFormat: "response_format",
        streamOptions: "stream_options",
        toolChoice: "tool_choice",
        topP: "top_p",
    });
});
export function chatGenerationParamsToJSON(chatGenerationParams) {
    return JSON.stringify(ChatGenerationParams$outboundSchema.parse(chatGenerationParams));
}
